# 协程

https://www.liaoxuefeng.com/wiki/1016959663602400/1017968846697824

一文中，对协程优势进行了描述：协程不需要多线程的锁机制，因为只有一个线程，也**不存在同时写变量冲突**，在协程中控制共享资源不加锁，只需要判断状态就好了，所以执行效率比多线程高很多。

我个人的问题是，为何不存在同时写变量冲突？

下面用一个案例试图说明下。

首先是多线程的案例，这里使用了4个线程。

## 多线程

```python
import threading
import time

n = 1000
def func():
    global n
    while n > 0:
        for _ in range(10000):
            continue
        n = n - 1

for _ in range(4):
    threading.Thread(target=func).start()

time.sleep(.5)
print("最终 n =", n)
```

最终结果往往是错误的：

```
最终 n = -3
```

为什么结果是错误的？

上面的案例中，用一个循环10000次的for循环模拟了一些耗时的操作

```python
for _ in range(10000):
    continue
```

而在多线程运行的过程中，代码的执行是类似分配时间片运行的，因此很可能在for循环执行过程中，切换到其他线程了，这就导致当n=1时进入了某个线程，但是还没来得及完成减1的语句，就切换到了其他线程，此时n还是1，这就导致n被重复减了。

## 协程

在多线程的条件下，代码执行顺序是无序的。

下面将代码改为协程

```python
import asyncio
import time

n = 1000
async def func():
    global n
    while n > 0:
        for _ in range(10000):
            continue
        n = n - 1

async def main():
    tasks = [asyncio.create_task(func()) for _ in range(4)]
    await asyncio.gather(*tasks)

asyncio.run(main())
time.sleep(.5)
print("最终 n =", n)
```

```
最终 n = 0
```

为什么不会错误？

在协程中，如果没有遇到`await`语句，是一定不会切换到其他协程的。

这就保证了代码的有序执行。虽然for循环循环了10000次，但是这期间代码是绝对不会切换到其他协程的。

从这个例子来看，协程和多线程最大的差别是：多线程可能在任何地方切换到其他线程，而协程什么时候切换由程序员决定。

比如下面这个例子，用`await asyncio.sleep(0.001)`代替了for循环，其他代码一样

```python
import asyncio
import time

n = 1000
async def func():
    global n
    while n > 0:
        for _ in range(10000):
            continue
        n = n - 1

async def main():
    tasks = [asyncio.create_task(func()) for _ in range(4)]
    await asyncio.gather(*tasks)

asyncio.run(main())
time.sleep(.5)
print("最终 n =", n)
```

```
最终 n = -3
```

这时，最终计算的结果又错误了。

这说明**协程切换时机不正确时，也会发生重复写**的情况。



关键是`while n>0`到`n=n-1`之间不要切换协程。

也就是这一段代码得全部用同步代码不要出现`await`，这样就不会造成错误了。

## 协程锁

https://docs.python.org/3/library/asyncio-sync.html?highlight=asyncio%20lock#asyncio.Lock

有些场景下，还是需要类似线程锁这样的协程锁。

可以参考下面的网址：

https://stackoverflow.com/questions/25799576/whats-python-asyncio-lock-for

比如说，在编写函数的过程中，有一个辅助函数是用协程写的。比如下面这个例子，有一个函数`do_something()`可能会导致切换到其他协程。但是还必须得用。

如果不改代码的话，就会在`n=n-1`之前切换出去，导致最终的n可能会小于0。

```python
async def do_something():
    await asyncio.sleep(0.001)
    
async def func():
    global n
    while n > 0:
        # 必须要用这个辅助函数，或者这个函数不方便改成同步代码
        await do_something()
        n = n - 1
```

这时候可以使用协程锁，协程锁的使用方式如下，首先声明一把锁，然后在需要的地方使用它。

被`async with lock:`包裹的代码块，虽然里边有`await`，但是会被保证为不切换出去，也就是这部分代码块变成同步代码了。

```python
lock = asyncio.Lock()
async def func():
    global n
    while n > 0:
        # 必须要用这个辅助函数，或者这个函数不方便改成同步代码
        async with lock:
            await do_something()
            n = n - 1
```

完整代码如下：

```python
import asyncio
import time

n = 1000
lock = asyncio.Lock()

async def do_something():
    await asyncio.sleep(0.001)
    
async def func():
    global n, lock
    while n > 0:
        # 必须要用这个辅助函数，或者这个函数不方便改成同步代码
        async with lock:
            await do_something()
            n = n - 1

async def main():
    tasks = [asyncio.create_task(func()) for _ in range(4)]
    await asyncio.gather(*tasks)

asyncio.run(main())
time.sleep(.5)
print("最终 n =", n)
```

ps：这个例子不太对，好像我还没找到合适的方法。

要避免这种问题，最方便的是，在真正赋值之前加个if语句判断一下。

```python
if n > 0:
    n = n - 1
```



## 总结

也就是说，协程可以自己控制什么时候切换协程。

对于一个协程函数，没有`await`的时候，代码是不会随意切换的。

比如对于下面这个代码段

```python
for _ in range(10000):
    continue
```

对于协程而言，一定不会在for循环中间切换出去，一直执行完整个for循环。

而对于多线程，可能在任意一次for循环中切换到其他线程。





